### PR TITLE
PM-5550 Implement on-page autofil for single line TOTP

### PR DESCRIPTION
**Pull request by dan-livefront**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-5550
## Core Changes

* Introduce a feature flag (  InlineMenuTotp = "inline-menu-totp" ).
* Update the disqualification logic to no longer exclude TOTP fields if the feature flag is enabled.

**Update Inline Menu Cipher Data**

Modify the inline menu cipher data to include the following fields:
- totpCode: The generated TOTP code.
- totpTimeInterval: Time left for the current TOTP code.
- totpField: Boolean indicating if the field is a TOTP field.

**Populate New Inline Menu for TOTP Fields**

When a TOTP code is available and the field is identified as a TOTP field:

- Display the TOTP code in the inline menu.
- Include a countdown timer (similar to the existing one in the extension) showing the remaining validity of the code.
- Add a heading explicitly labeling it as a TOTP code.
- When the totp code expires, refresh overlay ciphers

**Tests**

- Test the behavior with the feature flag enabled and disabled:
- Ensure TOTP fields are not disqualified when the flag is enabled.
- Confirm default behavior (TOTP fields disqualified) when the flag is disabled and enabled

## Note:

- I organized some of the styling file names to not directly assign a hex code in dark mode scenarios but I can remove these changes if they will add to much friction to this PR.

- [this ticket](https://bitwarden.atlassian.net/jira/software/c/projects/BW/boards/93?assignee=712020%3Abf2a72c1-ce10-4127-b0ee-9f39b6514ad8&selectedIssue=PM-15437&sprint=481)
This ticket will handle multiple totp login items on the same field

- [this ticket](https://bitwarden.atlassian.net/browse/PM-15189) This ticket will handle the way we are determining if a field is a totp field on the page 

- [this ticket]( https://bitwarden.atlassian.net/jira/software/c/projects/BW/boards/93?assignee=712020%3Abf2a72c1-ce10-4127-b0ee-9f39b6514ad8&assignee=712020%3A920ea2a8-c3ca-4d92-a441-3b85639f8101&selectedIssue=PM-14954)  handles multi input totp fields

- [this ticket](https://bitwarden.atlassian.net/browse/PM-15591 ) this ticket will mitigate showing totp codes when MP prompt is on

- [This ticket](https://bitwarden.atlassian.net/browse/PM-15592) adds the subscription paywall

- [This ticket](https://bitwarden.atlassian.net/browse/PM-15593) will remove totp code inline menu items that do not have a totp code

## 📸 Screenshots

Screenshot 2024-11-19 at 3 37 49 PM
Screenshot 2024-11-19 at 3 37 55 PM

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

**Review by danielleflinn — APPROVED** [review]
Tested build locally, UI looks as expected and overall works as expected when navigating with a mouse as well as with a screen reader and keyboard. 

I did catch 2 bugs / edge cases that we hadn't thought of:

1. We are showing the TOTP fill UI when an item has MP reprompt "on" which feels a bit odd since selecting it to autofill then launches the popout. And in the Extension Refresh we actually require the user to enter their MP to view a vault item OR to copy the TOTP. 

2. Secondly, for Free users we are showing the TOTP menu UI even though we are not showing the TOTP in the popup (requires premium upgrade). This was an edge case that did not come up during design. 

I'll likely log both of these as bugs on the Jira though so they can be addressed in separate PRs.

File: apps/browser/src/autofill/overlay/inline-menu/pages/list/autofill-inline-menu-list.ts [hunk 1]
```diff
@@ -1044,6 +1044,66 @@ export class AutofillInlineMenuList extends AutofillInlineMenuPageElement {
     cipherIcon.classList.add("cipher-icon");
     cipherIcon.setAttribute("aria-hidden", "true");
 
+    if (cipher.login?.totpField && cipher.login?.totp) {
+      const totpContainer = document.createElement("div");
+      totpContainer.style.position = "relative";
+
+      const svgElement = buildSvgDomElement(`
+      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 29 29">
+        <g>
```
Comment (jprusik) [inline apps/browser/src/autofill/overlay/inline-menu/pages/list/autofill-inline-menu-list.ts:1053]: non-blocking nit; are we able to drop this group tag from the SVG?

File: apps/browser/src/autofill/overlay/inline-menu/pages/list/autofill-inline-menu-list.ts [hunk 2]
```diff
@@ -1044,6 +1044,66 @@ export class AutofillInlineMenuList extends AutofillInlineMenuPageElement {
...
+        /**
+         * Design specifies a seven-second time span as the period where expiry is approaching.
+         */
+        const totpExpiryApproaching = totpSeconds <= 7;
+
+        totpSecondsSpan.classList.toggle("totp-sec-span-danger", totpExpiryApproaching);
+        innerCircleElement.classList.toggle("circle-danger-color", totpExpiryApproaching);
+        outerCircleElement.classList.toggle("circle-danger-color", totpExpiryApproaching);
+
+        innerCircleElement.style.strokeDashoffset = `${((intervalSeconds - totpSeconds) / intervalSeconds) * (2 * Math.PI * 12.5)}`;
```
Comment (jprusik) [inline apps/browser/src/autofill/overlay/inline-menu/pages/list/autofill-inline-menu-list.ts:1094]: ✨

**Comment by jprusik** [comment]
BIT run:
https://github.com/bitwarden/browser-interactions-testing/actions/runs/12184082729

File: apps/browser/src/_locales/en/messages.json [hunk 3]
```diff
@@ -3574,6 +3581,15 @@
...
   },
+  "totpCodeAria":{
+    "message": "Time-based One-Time Password Verification Code",
+    "description":"Aria label for the totp code displayed in the inline menu for autofill"
+  },
+  "totpSecondsSpanAria":{
+    "message":"Time remaining before current TOTP expires",
+    "description":"Aria label for the totp seconds displayed in the inline menu for autofill"
+
+  },
```
Comment (jprusik) [inline apps/browser/src/_locales/en/messages.json:3592]: nits/cleanup

```suggestion
  "totpCodeAria": {
    "message": "Time-based One-Time Password Verification Code",
    "description": "Aria label for the totp code displayed in the inline menu for autofill"
  },
  "totpSecondsSpanAria": {
    "message": "Time remaining before current TOTP expires",
    "description": "Aria label for the totp seconds displayed in the inline menu for autofill"
  },
```

File: apps/browser/src/_locales/en/messages.json [hunk 4]
```diff
@@ -192,6 +192,13 @@
   "autoFillIdentity": {
     "message": "Autofill identity"
   },
+  "fillVerificationCode":{
+    "message": "Fill verification code"
+  },
+  "fillVerificationCodeAria":{
+    "message": "Fill Verification Code",
+    "description":"Aria label for the heading displayed the inline menu for totp code autofill"
+  },
```
Comment (jprusik) [inline apps/browser/src/_locales/en/messages.json:201]: formatting nits

```suggestion
  "fillVerificationCode": {
    "message": "Fill verification code"
  },
  "fillVerificationCodeAria": {
    "message": "Fill Verification Code",
    "description": "Aria label for the heading displayed the inline menu for totp code autofill"
  },
```

File: apps/browser/src/autofill/overlay/inline-menu/pages/list/autofill-inline-menu-list.ts (hunk 1 above)
Comment (dan-livefront) [reply apps/browser/src/autofill/overlay/inline-menu/pages/list/autofill-inline-menu-list.ts:1053]: The group can be dropped and won't affect any functionality, I will drop it.

Comment (dan-livefront) [reply apps/browser/src/autofill/overlay/inline-menu/pages/list/autofill-inline-menu-list.ts]: [Tag removed ](https://github.com/bitwarden/clients/pull/12058/commits/c9831a425c5703402e7e608e65ffbdf259018e20)

File: apps/browser/src/background/runtime.background.ts [hunk 5]
```diff
@@ -71,6 +71,7 @@ export default class RuntimeBackground {
         "biometricUnlockAvailable",
         "getUseTreeWalkerApiForPageDetailsCollectionFeatureFlag",
         "getInlineMenuFieldQualificationFeatureFlag",
+        "getInlineMenuTotpFlag",
```
Comment (jprusik) [inline apps/browser/src/background/runtime.background.ts:74]:
```suggestion
        "getInlineMenuTotpFeatureFlag",
```

File: apps/browser/src/autofill/services/inline-menu-field-qualification.service.ts [hunk 6]
```diff
@@ -167,8 +171,11 @@ export class InlineMenuFieldQualificationService
       return this.isFieldForLoginFormFallback(field);
     }
 
-    if (this.isTotpField(field)) {
-      return false;
+    const isTotpField = this.isTotpField(field);
+    if (this.inlineMenuTotpFeatureFlag) {
+      if (isTotpField) {
+        return true;
+      }
```
Comment (jprusik) [inline apps/browser/src/autofill/services/inline-menu-field-qualification.service.ts:178]: nit

```suggestion
    if (this.inlineMenuTotpFeatureFlag) {
      const isTotpField = this.isTotpField(field);
      
      if (isTotpField) {
        return true;
      }
```

Comment (dan-livefront) [reply apps/browser/src/autofill/services/inline-menu-field-qualification.service.ts]: [addressed here](https://github.com/bitwarden/clients/pull/12058/commits/9c2c03ee38f27fe8c9e178a73ea0948e6264e613)

File: apps/browser/src/background/runtime.background.ts (hunk 5 above)
Comment (dan-livefront) [reply apps/browser/src/background/runtime.background.ts]: nice catch, Jon. [addressed here](https://github.com/bitwarden/clients/pull/12058/commits/9c2c03ee38f27fe8c9e178a73ea0948e6264e613)

**Review by jprusik — DISMISSED** [review]
v. nice work; thanks @dan-livefront

**Review by jprusik — APPROVED** [review]

File: apps/browser/src/autofill/services/inline-menu-field-qualification.service.ts [hunk 7]
```diff
@@ -171,9 +171,13 @@ export class InlineMenuFieldQualificationService
...
 
-    const isTotpField = this.isTotpField(field);
+    /**
+     * Autofill does not fill password type totp input fields
+     */
     if (this.inlineMenuTotpFeatureFlag) {
-      if (isTotpField) {
+      const isTotpField = this.isTotpField(field);
+      const passwordType = field.type === "password";
+      if (isTotpField && !passwordType) {
```
[the remark this answers is not in the document]
Comment (dan-livefront) [reply apps/browser/src/autofill/services/inline-menu-field-qualification.service.ts]: Will do! good call out.

**Comment by jprusik** [comment]
New BIT run with the "inline-menu-field-qualification" flag enabled as well:
https://github.com/bitwarden/browser-interactions-testing/actions/runs/12299952000

**Pull request merged**